File: py/google/fhir/references.py
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Utilities for traversing and manipulating reference types."""

import re
from typing import Optional

from google.protobuf import descriptor
from google.protobuf import message
from google.fhir.utils import annotation_utils
from google.fhir.utils import path_utils
from google.fhir.utils import proto_utils

_FRAGMENT_REFERENCE_PATTERN = re.compile(r'^#[A-Za-z0-9.-]{1,64}$')
_INTERNAL_REFERENCE_PATTERN = re.compile(
    r'^(?P<resource_type>[0-9A-Za-z_]+)/(?P<resource_id>[A-Za-z0-9.-]{1,64})'
    r'(?:/_history/(?P<version>[A-Za-z0-9.-]{1,64}))?$')


def _validate_reference(reference: message.Message):
  """Raises a ValueError if the provided Message is not a FHIR reference."""
  if not annotation_utils.is_reference(reference):
    raise ValueError(
        f'Message {reference.DESCRIPTOR.name} is not a FHIR reference.')


def get_reference_id_field_for_resource(
    reference: message.Message,
    resource_type: str) -> descriptor.FieldDescriptor:
  """Returns the reference ID field for a provided resource type."""
  _validate_reference(reference)
  field_name = path_utils.camel_case_to_snake_case(resource_type) + '_id'
  field = reference.DESCRIPTOR.fields_by_name.get(field_name)
  if field is None:
    raise ValueError(f'Resource type {resource_type!r} is not valid for a '
                     f'reference. Field {field_name!r} does not exist.')
  return field


def populate_typed_reference_id(reference_id: message.Message, resource_id: str,
                                version: Optional[str]):
  """Sets the resource_id and optionally, version, on the reference."""
  reference_id_value_field = reference_id.DESCRIPTOR.fields_by_name['value']
  proto_utils.set_value_at_field(reference_id, reference_id_value_field,
                                 resource_id)
  if version is not None:
    history_field = reference_id.DESCRIPTOR.fields_by_name.get('history')
    if history_field is None:
      raise ValueError('Not a valid ReferenceId message: '
                       f"{reference_id.DESCRIPTOR.full_name}. Field 'history' "
                       'does not exist.')
    history = proto_utils.set_in_parent_or_add(reference_id, history_field)
    history_value_field = history.DESCRIPTOR.fields_by_name['value']
    proto_utils.set_value_at_field(history, history_value_field, version)


def split_if_relative_reference(reference: message.Message):
  """If possible, parses a `Reference` `uri` into more structured fields.

  This is only possible for two forms of reference uris:
  * Relative references of the form $TYPE/$ID, e.g., "Patient/1234"
    In this case, this will be parsed to a proto of the form:
    {patient_id: {value: "1234"}}
  * Fragments of the form "#$FRAGMENT", e.g., "#vs1".  In this case, this would
    be parsed into a proto of the form:
    {fragment: {value: "vs1"} }

  If the reference URI matches one of these schemas, the `uri` field will be
  cleared, and the appropriate structured fields set. Otherwise, the reference
  will be unchanged.

  Args:
    reference: The FHIR reference to potentially split.

  Raises:
    ValueError: If the message is not a valid FHIR Reference proto.
  """
  _validate_reference(reference)
  uri_field = reference.DESCRIPTOR.fields_by_name.get('uri')
  if not proto_utils.field_is_set(reference, uri_field):
    return  # No URI to split

  uri = proto_utils.get_value_at_field(reference, uri_field)

  internal_match = re.fullmatch(_INTERNAL_REFERENCE_PATTERN, uri.value)
  if internal_match is not None:
    # Note that we make the reference_id off of the reference before adding it,
    # since adding the reference_id would destroy the uri field, as they are
    # both in the same oneof. This allows us to copy fields from uri to
    # reference_id without making an extra copy.
    reference_id_field = get_reference_id_field_for_resource(
        reference, internal_match.group('resource_type'))
    reference_id = proto_utils.create_message_from_descriptor(
        reference_id_field.message_type)
    populate_typed_reference_id(reference_id,
                                internal_match.group('resource_id'),
                                internal_match.group('version'))

    proto_utils.copy_common_field(uri, reference_id, 'id')
    proto_utils.copy_common_field(uri, reference_id, 'extension')
    proto_utils.set_value_at_field(reference, reference_id_field, reference_id)
    return

  fragment_match = re.fullmatch(_FRAGMENT_REFERENCE_PATTERN, uri.value)
  if fragment_match is not None:
    # Note that we make the fragment off of the reference before adding it,
    # since adding the fragment would destroy the uri field, as they are both in
    # the same oneof. This allows us to copy fields from uri to fragment without
    # making an extra copy.
    fragment_field = reference.DESCRIPTOR.fields_by_name['fragment']
    fragment = proto_utils.create_message_from_descriptor(
        fragment_field.message_type)

    value_field = fragment.DESCRIPTOR.fields_by_name['value']
    proto_utils.set_value_at_field(fragment, value_field, uri.value[1:])
    proto_utils.copy_common_field(uri, fragment, 'id')
    proto_utils.copy_common_field(uri, fragment, 'extension')
    proto_utils.set_value_at_field(reference, fragment_field, fragment)
    return


def reference_to_string(reference: message.Message) -> str:
  """Returns a reference URI for a typed reference message."""
  _validate_reference(reference)

  # Early-exit if URI or fragment is set
  if proto_utils.field_is_set(reference, 'uri'):
    uri = proto_utils.get_value_at_field(reference, 'uri')
    return uri.value
  elif proto_utils.field_is_set(reference, 'fragment'):
    fragment = proto_utils.get_value_at_field(reference, 'fragment')
    return f'#{fragment.value}'

  set_oneof = reference.WhichOneof('reference')
  if set_oneof is None:
    raise ValueError(f'Reference is not set on: {reference.DESCRIPTOR.name}.')

  # Convert to CamelCase
  prefix = path_utils.snake_case_to_camel_case(set_oneof, upper=True)

  # Cull trailing 'Id'
  if prefix.endswith('Id'):
    prefix = prefix[:-2]

  reference_id = proto_utils.get_value_at_field(reference, set_oneof)
  reference_string = f'{prefix}/{reference_id.value}'
  if proto_utils.field_is_set(reference_id, 'history'):
    reference_string += f'/_history/{reference_id.history.value}'
  return reference_string

File: cc/google/fhir/references.cc
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "google/fhir/references.h"

#include <memory>
#include <string>

#include "absl/memory/memory.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "absl/types/optional.h"
#include "google/fhir/proto_util.h"
#include "google/fhir/status/status.h"
#include "google/fhir/status/statusor.h"
#include "google/fhir/util.h"
#include "google/protobuf/descriptor.h"
#include "google/protobuf/message.h"
#include "re2/re2.h"

namespace google {
namespace fhir {

using ::absl::InvalidArgumentError;
using ::google::protobuf::Descriptor;
using ::google::protobuf::FieldDescriptor;
using ::google::protobuf::Message;
using ::google::protobuf::Reflection;

absl::Status ReferenceStringToProto(const std::string& input,
                                    Message* reference) {
  const Descriptor* descriptor = reference->GetDescriptor();
  const Reflection* reflection = reference->GetReflection();
  const FieldDescriptor* uri_field = descriptor->FindFieldByName("uri");
  FHIR_RETURN_IF_ERROR(SetPrimitiveStringValue(
      reflection->MutableMessage(reference, uri_field), input));
  return SplitIfRelativeReference(reference);
}

std::string GetReferenceStringToResource(const Message& message) {
  return absl::StrCat(message.GetDescriptor()->name(), "/",
                      GetResourceId(message).value());
}

absl::StatusOr<absl::optional<std::string>> ReferenceProtoToString(
    const ::google::protobuf::Message& reference) {
  const ::google::protobuf::Descriptor* descriptor = reference.GetDescriptor();
  const ::google::protobuf::Reflection* reflection = reference.GetReflection();

  std::string uri_value;
  FHIR_ASSIGN_OR_RETURN(uri_value,
                        GetPrimitiveStringValue(reference, "uri", &uri_value));
  if (!uri_value.empty()) {
    return uri_value;
  }

  std::string fragment_value;
  FHIR_ASSIGN_OR_RETURN(
      fragment_value,
      GetPrimitiveStringValue(reference, "fragment", &fragment_value));
  if (!fragment_value.empty()) {
    return absl::StrCat("#", fragment_value);
  }

  static const google::protobuf::OneofDescriptor* oneof =
      descriptor->FindOneofByName("reference");
  if (oneof == nullptr) {
    return absl::InvalidArgumentError(
        absl::StrCat("Invalid Reference type: ", descriptor->full_name()));
  }
  const ::google::protobuf::FieldDescriptor* reference_field =
      reflection->GetOneofFieldDescriptor(reference, oneof);
  if (reference_field == nullptr) {
    return absl::nullopt;
  }
  if (reference_field->name() == "uri" ||
      reference_field->name() == "fragment") {
    return absl::nullopt;
  }

  std::string prefix;
  bool start = true;
  for (const char c : reference_field->name()) {
    if (start) {
      start = false;
      prefix.push_back(c + 'A' - 'a');
    } else if (c == '_') {
      start = true;
    } else {
      prefix.push_back(c);
    }
  }
  static LazyRE2 re = {"Id$"};
  RE2::Replace(&prefix, *re, "");

  const ::google::protobuf::Message& id =
      reflection->GetMessage(reference, reference_field);

  std::string id_value;
  FHIR_ASSIGN_OR_RETURN(id_value, GetPrimitiveStringValue(id, &id_value));
  std::string reference_string = absl::StrCat(prefix, "/", id_value);

  std::string history_value;
  FHIR_ASSIGN_OR_RETURN(history_value,
                        GetPrimitiveStringValue(id, "history", &history_value));

  if (!history_value.empty()) {
    absl::StrAppend(&reference_string, "/_history/", history_value);
  }
  return reference_string;
}

// Splits relative references into their components, for example, "Patient/ABCD"
// will result in the patientId field getting the value "ABCD".
absl::Status SplitIfRelativeReference(Message* reference) {
  const Descriptor* descriptor = reference->GetDescriptor();
  const Reflection* reflection = reference->GetReflection();

  const FieldDescriptor* uri_field = descriptor->FindFieldByName("uri");
  const FieldDescriptor* fragment_field =
      descriptor->FindFieldByName("fragment");
  if (!reflection->HasField(*reference, uri_field)) {
    // There is no uri to split
    return absl::OkStatus();
  }

  const Message& uri = reflection->GetMessage(*reference, uri_field);

  std::string uri_scratch;
  FHIR_ASSIGN_OR_RETURN(const std::string& uri_string,
                        GetPrimitiveStringValue(uri, &uri_scratch));

  static const LazyRE2 kInternalReferenceRegex{
      "([0-9A-Za-z_]+)/([A-Za-z0-9.-]{1,64})(?:/_history/"
      "([A-Za-z0-9.-]{1,64}))?"};
  std::string resource_type;
  std::string resource_id;
  std::string version;
  if (RE2::FullMatch(uri_string, *kInternalReferenceRegex, &resource_type,
                     &resource_id, &version)) {
    const FieldDescriptor* reference_id_field =
        internal::GetReferenceFieldForResource(*reference, resource_type);
    if (reference_id_field == nullptr) {
      // Not a recognized relative reference.
      return absl::OkStatus();
    }

    // Note that we make the reference_id off of the reference before adding it,
    // since adding the reference_id would destroy the uri field, since they are
    // in the same oneof.  This way allows us to copy fields from uri to
    // reference_id without an extra copy.
    std::unique_ptr<Message> reference_id =
        absl::WrapUnique(reflection->GetMessageFactory()
                             ->GetPrototype(reference_id_field->message_type())
                             ->New());
    FHIR_RETURN_IF_ERROR(internal::PopulateTypedReferenceId(
        resource_id, version, reference_id.get()));
    const Message& uri = reflection->GetMessage(*reference, uri_field);
    reference_id->GetTypeName();
    FHIR_RETURN_IF_ERROR(CopyCommonField(uri, reference_id.get(), "id"));
    FHIR_RETURN_IF_ERROR(CopyCommonField(uri, reference_id.get(), "extension"));
    reflection->SetAllocatedMessage(reference, reference_id.release(),
                                    reference_id_field);
    return absl::OkStatus();
  }

  static const LazyRE2 kFragmentReferenceRegex{"#[A-Za-z0-9.-]{1,64}"};
  if (RE2::FullMatch(uri_string, *kFragmentReferenceRegex)) {
    // Note that we make the fragment off of the reference before adding it,
    // since adding the fragment would destroy the uri field, since they are in
    // the same oneof.  This way allows us to copy fields from uri to fragment
    // without an extra copy.
    std::unique_ptr<Message> fragment =
        absl::WrapUnique(reflection->GetMessageFactory()
                             ->GetPrototype(fragment_field->message_type())
                             ->New());
    FHIR_RETURN_IF_ERROR(
        SetPrimitiveStringValue(fragment.get(), uri_string.substr(1)));
    FHIR_RETURN_IF_ERROR(CopyCommonField(uri, fragment.get(), "id"));
    FHIR_RETURN_IF_ERROR(CopyCommonField(uri, fragment.get(), "extension"));
    reflection->SetAllocatedMessage(reference, fragment.release(),
                                    fragment_field);
    return absl::OkStatus();
  }

  // There's no way to rewrite the URI, but it's valid as is.
  return absl::OkStatus();
}

namespace internal {

absl::Status PopulateTypedReferenceId(const std::string& resource_id,
                                      const std::string& version,
                                      Message* reference_id) {
  FHIR_RETURN_IF_ERROR(SetPrimitiveStringValue(reference_id, resource_id));
  if (!version.empty()) {
    const FieldDescriptor* history_field =
        reference_id->GetDescriptor()->FindFieldByName("history");
    if (history_field == nullptr) {
      return InvalidArgumentError(
          absl::StrCat("Not a valid ReferenceId message: ",
                       reference_id->GetDescriptor()->full_name(),
                       ".  Field history does not exist)."));
    }
    FHIR_RETURN_IF_ERROR(
        SetPrimitiveStringValue(reference_id->GetReflection()->MutableMessage(
                                    reference_id, history_field),
                                version));
  }
  return absl::OkStatus();
}

const FieldDescriptor* GetReferenceFieldForResource(
    const Message& reference, absl::string_view resource_type) {
  const std::string field_name =
      absl::StrCat(ToSnakeCase(resource_type), "_id");
  return reference.GetDescriptor()->FindFieldByName(field_name);
}

}  // namespace internal

}  // namespace fhir
}  // namespace google
